src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -360,6 +360,60 @@ public static ImmutableArray<ParameterSymbol> DelegateParameters(this TypeSymbol
             return type.DelegateInvokeMethod().Parameters;
         }
 
+        public static bool TryGetElementTypesIfTupleOrCompatible(this TypeSymbol type, out ImmutableArray<TypeSymbol> elementTypes)
+        {
+            if (type.IsTupleType)
+            {
+                elementTypes = ((TupleTypeSymbol)type).TupleElementTypes;
+                return true;
+            }
+
+            // The following codepath should be very uncommon since it would be rare
+            // to see a tuple underlying type not represented as a tuple.
+            // It still might happen since tuple underlying types are creatable via public APIs 
+            // and it is also possible that they would be passed in.
+
+            // PERF: if allocations here become nuisance, consider caching the results
+            //       in the type symbols that can actually be tuple compatible
+            int cardinality;
+            if (!type.IsTupleCompatible(out cardinality))
+            {
+                // source not a tuple or compatible
+                elementTypes = default(ImmutableArray<TypeSymbol>);
+                return false;
+            }
+
+            var elementTypesBuilder = ArrayBuilder<TypeSymbol>.GetInstance(cardinality);
+            TupleTypeSymbol.AddElementTypes((NamedTypeSymbol)type, elementTypesBuilder);
+
+            Debug.Assert(elementTypesBuilder.Count == cardinality);
+
+            elementTypes = elementTypesBuilder.ToImmutableAndFree();
+            return true;
+        }
+
+       public static ImmutableArray<TypeSymbol> GetElementTypesIfTupleOrCompatible(this TypeSymbol type)
```

Comment:
Then perhaps we can share portion of the code among the two methods by extracting it in a private helper method.

---
In reply to: [66329168](https://github.com/dotnet/roslyn/pull/11681#discussion_r66329168) [](ancestors = 66329168)